python/tritonserver/_c/tritonserver_pybind.cc

```diff
@@ -0,0 +1,2106 @@
+  }
+
+  std::shared_ptr<PyMessage> ModelConfig(
+      const std::string& model_name, int64_t model_version,
+      uint32_t config_version = 1) const
+  {
+    struct TRITONSERVER_Message* model_config;
+    ThrowIfError(TRITONSERVER_ServerModelConfig(
+        triton_object_, model_name.c_str(), model_version, config_version,
+        &model_config));
+    return std::make_shared<PyMessage>(model_config, true /* owned */);
+  }
+
+  std::shared_ptr<PyMessage> ModelIndex(uint32_t flags) const
+  {
+    struct TRITONSERVER_Message* model_index;
+    ThrowIfError(
+        TRITONSERVER_ServerModelIndex(triton_object_, flags, &model_index));
+    return std::make_shared<PyMessage>(model_index, true /* owned */);
+  }
+
+  void LoadModel(const std::string& model_name)
+  {
+    // load model is blocking, ensure to release GIL
+    py::gil_scoped_release release;
+    ThrowIfError(
+        TRITONSERVER_ServerLoadModel(triton_object_, model_name.c_str()));
+  }
+
+  void LoadModelWithParameters(
+      const std::string& model_name,
+      const std::vector<std::shared_ptr<PyParameter>>& parameters) const
+  {
+    std::vector<const struct TRITONSERVER_Parameter*> params;
+    for (const auto& p : parameters) {
+      params.emplace_back(p->Ptr());
+    }
+    // load model is blocking, ensure to release GIL
+    py::gil_scoped_release release;
+    ThrowIfError(TRITONSERVER_ServerLoadModelWithParameters(
+        triton_object_, model_name.c_str(), params.data(), params.size()));
+  }
+
+  void UnloadModel(const std::string& model_name)
+  {
+    ThrowIfError(
+        TRITONSERVER_ServerUnloadModel(triton_object_, model_name.c_str()));
+  }
+
+  void UnloadModelAndDependents(const std::string& model_name)
+  {
+    ThrowIfError(TRITONSERVER_ServerUnloadModelAndDependents(
+        triton_object_, model_name.c_str()));
+  }
+
+  std::shared_ptr<PyMetrics> Metrics() const
+  {
+    struct TRITONSERVER_Metrics* metrics;
+    ThrowIfError(TRITONSERVER_ServerMetrics(triton_object_, &metrics));
+    return std::make_shared<PyMetrics>(metrics, true /* owned */);
+  }
+
+  void InferAsync(
+      const std::shared_ptr<PyInferenceRequest>& request, PyTrace& trace)
+  {
+    // Extra handling to avoid circular inclusion:
+    //   request -> request_callback_resource_ -> request
+    // 1. extract 'request_callback_resource_' out and provide
+    //    scoped handler to place resource back to request if not released,
+    //    TRITONSERVER_ServerInferAsync failed in other words.
+    // 2. add 'request' into resource so request release callback can access it.
+    // 3. call TRITONSERVER_ServerInferAsync.
+    // 4. release the extracted resource if TRITONSERVER_ServerInferAsync
+    //    returns.
+    static auto resource_handler =
+        [](PyInferenceRequest::CallbackResource* cr) {
+          if (cr != nullptr) {
+            cr->request->request_callback_resource_.reset(cr);
+            cr->request.reset();
+          }
```

Comment:
Deleting `cr` is not required because it will be deleted by the `request` object?

Reply:
Correct, the above smart pointer ops are to move `cr` ownership back to `request`